sharpd/sharp_nht.c

```diff
@@ -65,3 +69,149 @@ void sharp_nh_tracker_dump(struct vty *vty)
 			nht->updates);
 	}
 }
+
+PREDECL_RBTREE_UNIQ(sharp_nhg_rb);
+
+struct sharp_nhg {
+	struct sharp_nhg_rb_item mylistitem;
+
+	uint32_t id;
+
+	char name[256];
+
+	bool installed;
+};
+
+static uint32_t nhg_id;
+
+static uint32_t sharp_get_next_nhid(void)
+{
+	zlog_debug("Id assigned: %u", nhg_id);
+	return nhg_id++;
+}
+
+struct sharp_nhg_rb_head nhg_head;
+
+static int sharp_nhg_compare_func(const struct sharp_nhg *a,
+				  const struct sharp_nhg *b)
+{
+	return strncmp(a->name, b->name, strlen(a->name));
+}
+
+DECLARE_RBTREE_UNIQ(sharp_nhg_rb, struct sharp_nhg, mylistitem,
+		    sharp_nhg_compare_func);
+
+static struct sharp_nhg *sharp_nhgroup_find_id(uint32_t id)
+{
+	struct sharp_nhg *lookup;
+
+	/* Yea its just a for loop, I don't want add complexity
+	 * to sharpd with another RB tree for just IDs
+	 */
+
+	frr_each (sharp_nhg_rb, &nhg_head, lookup) {
+		if (lookup->id == id)
+			return lookup;
+	}
+
+	return NULL;
+}
+
+static void sharp_nhgroup_add_cb(const char *name)
```

Comment:
so ... this looks like every nhg will get an id assigned to it - is that right? If so, it looks as if that will force all sharp use of nhgs to use the ids only, and that will break things that ... other folks use (like backups?)

Reply:
When you create the nexthop-group, it will send it to zebra. If it fails to install it because its not a type we support or for some other reason, sharpd gets told and reverts to just using the traditional way of sending all nexthops down with the route, rather than ID.